pkg/auth/oauth/external/gitlab/gitlab.go

```diff
@@ -0,0 +1,111 @@
+package gitlab
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"github.com/RangelReale/osincli"
+	"github.com/golang/glog"
+
+	authapi "github.com/openshift/origin/pkg/auth/api"
+	"github.com/openshift/origin/pkg/auth/oauth/external"
+)
+
+const (
+	gitlabAuthorizePath = "oauth/authorize"
+	gitlabTokenPath     = "oauth/token"
+	gitlabUserApiPath   = "api/v3/user"
+	gitlabOAuthScope    = "api"
+)
+
+type provider struct {
+	providerName, hostURL, clientID, clientSecret string
+}
+
+type gitlabUser struct {
+	ID       uint64
+	Username string
+	Email    string
+	Name     string
+}
+
+func NewProvider(providerName, hostURL, clientID, clientSecret string) (external.Provider, error) {
```

Comment:
needs to take a transport in cases where they're running against an internal gitlab deployment and need to override the CA